File: src/com/colacelli/ircbot/plugins/access/AccessPlugin.java
package com.colacelli.ircbot.plugins.access;

import com.colacelli.ircbot.IRCBot;
import com.colacelli.ircbot.Plugin;
import com.colacelli.ircbot.listeners.OnChannelCommandListener;
import com.colacelli.ircbot.plugins.help.PluginHelp;
import com.colacelli.ircbot.plugins.help.PluginHelper;
import com.colacelli.irclib.connection.Connection;
import com.colacelli.irclib.messages.ChannelMessage;
import com.colacelli.irclib.messages.PrivateNoticeMessage;

public class AccessPlugin implements Plugin {
    @Override
    public String getName() {
        return "ACCESS";
    }

    @Override
    public void onLoad(IRCBot bot) {
        IRCBotAccess.getInstance().addListener(bot, IRCBotAccess.SUPER_ADMIN_LEVEL, new OnChannelCommandListener() {
            @Override
            public String channelCommand() {
                return ".access";
            }

            @Override
            public void onChannelCommand(Connection connection, ChannelMessage message, String command, String... args) {
                if (args != null) {
                    PrivateNoticeMessage.Builder builder = new PrivateNoticeMessage.Builder();
                    builder
                            .setSender(connection.getUser())
                            .setReceiver(message.getSender());

                    if (args.length > 1) {
                        switch (args[0]) {
                            case "add":
                                try {
                                    int level = Integer.parseInt(args[2]);
                                    IRCBotAccess.getInstance().setLevel(args[1], level);
                                    builder.setText("Access granted to " + args[1]);
                                } catch (NumberFormatException e) {
                                    builder.setText("Invalid level");
                                    // Invalid level
                                }
                                connection.send(builder.build());
                                break;

                            case "del":
                                IRCBotAccess.getInstance().setLevel(args[1], 0);
                                builder.setText("Access revoked to " + args[1]);
                                connection.send(builder.build());
                                break;
                        }
                    } else {
                        switch (args[0]) {
                            case "list":
                                IRCBotAccess.getInstance().getAccesses().forEach((nick, level) -> {
                                    builder.setText(nick + ": " + level);
                                    connection.send(builder.build());
                                });
                                break;
                        }
                    }
                }
            }
        });

        PluginHelper.getInstance().addHelp(new PluginHelp(
                ".access add",
                IRCBotAccess.SUPER_ADMIN_LEVEL,
                "Grant user access",
                "user",
                "level"));

        PluginHelper.getInstance().addHelp(new PluginHelp(
                ".access del",
                IRCBotAccess.SUPER_ADMIN_LEVEL,
                "Revoke user access",
                "user"));

        PluginHelper.getInstance().addHelp(new PluginHelp(
                ".access list",
                IRCBotAccess.SUPER_ADMIN_LEVEL,
                "Show access list"));
    }

    @Override
    public void onUnload(IRCBot bot) {
        IRCBotAccess.getInstance().removeListener(bot, ".access");
        PluginHelper.getInstance().removeHelp(".access add");
        PluginHelper.getInstance().removeHelp(".access del");
        PluginHelper.getInstance().removeHelp(".access list");
    }
}

File: src/com/colacelli/ircbot/plugins/access/AccessPlugin.kt
package com.colacelli.ircbot.plugins.access

import com.colacelli.ircbot.IRCBot
import com.colacelli.ircbot.base.Access
import com.colacelli.ircbot.base.Plugin
import com.colacelli.ircbot.base.listeners.OnChannelCommandListener
import com.colacelli.ircbot.base.Help
import com.colacelli.irclib.connection.Connection
import com.colacelli.irclib.messages.ChannelMessage
import com.colacelli.irclib.messages.PrivateNoticeMessage

class AccessPlugin : Plugin {
    override var name = "access"

    override fun onLoad(bot: IRCBot) {
        bot.addListener(object : OnChannelCommandListener {
            override var command = ".accessAdd"
            override var level = Access.Level.ROOT
            override var help = Help("Grant user access", "nick", "value")

            override fun onChannelCommand(connection: Connection, message: ChannelMessage, command: String, args: Array<String>) {
                if (args.size == 2) {
                    try {
                        val nick = args[0]
                        val level = Access.Level.valueOf(args[1].toUpperCase())

                        bot.access.add(nick, level)
                        connection.send(PrivateNoticeMessage("Access granted!", connection.user, message.sender))
                    } catch (e : IllegalArgumentException) {
                        connection.send(PrivateNoticeMessage("Invalid access value!", connection.user, message.sender))
                    }
                }
            }
        })

        bot.addListener(object : OnChannelCommandListener {
            override var command = ".accessDel"
            override var level = Access.Level.ROOT
            override var help = Help("Revoke user access", "nick")

            override fun onChannelCommand(connection: Connection, message: ChannelMessage, command: String, args: Array<String>) {
                if (args.isNotEmpty()) {
                    val nick = args[0]
                    bot.access.del(nick)
                    connection.send(PrivateNoticeMessage("Access revoked!", connection.user, message.sender))
                }
            }
        })

        bot.addListener(object : OnChannelCommandListener {
            override var command = ".accessList"

            override var level = Access.Level.ROOT

            override var help = Help("List user accesses")

            override fun onChannelCommand(connection: Connection, message: ChannelMessage, command: String, args: Array<String>) {
                var accesses = ""
                bot.access.list().forEach {
                    accesses += "${it.key}(${it.value.toString().toLowerCase()}) "
                }

                connection.send(PrivateNoticeMessage(accesses, connection.user, message.sender))
            }
        })
    }

    override fun onUnload(bot: IRCBot) {
        bot.removeListeners(arrayOf(".accessAdd", ".accessDel", ".accessList"))
    }
}